Description: Avoid sigpipe by first attempt to use send()
 Use send() then fall back to write in case of ENOTSOCK.
Author: Soren Hansen <soren@ubuntu.com>
Bug-Ubuntu: https://launchpad.net/bugs/718044
Index: plymouth-0.9.2/src/libply/ply-utils.c
===================================================================
--- plymouth-0.9.2.orig/src/libply/ply-utils.c
+++ plymouth-0.9.2/src/libply/ply-utils.c
@@ -274,6 +274,7 @@ ply_write (int         fd,
 {
         size_t bytes_left_to_write;
         size_t total_bytes_written = 0;
+        int is_a_socket = 1;
 
         assert (fd >= 0);
 
@@ -282,13 +283,26 @@ ply_write (int         fd,
         do {
                 ssize_t bytes_written = 0;
 
-                bytes_written = write (fd,
-                                       ((uint8_t *) buffer) + total_bytes_written,
-                                       bytes_left_to_write);
+                if (is_a_socket) {
+                        bytes_written = send  (fd,
+                               ((uint8_t *) buffer) + total_bytes_written,
+                               bytes_left_to_write,
+                               0
+#if HAVE_DECL_MSG_NOSIGNAL
+                               |MSG_NOSIGNAL
+#endif
+                               );
+                } else {
+                        bytes_written = write (fd,
+                                               ((uint8_t *) buffer) + total_bytes_written,
+                                               bytes_left_to_write);
+                }
 
                 if (bytes_written > 0) {
                         total_bytes_written += bytes_written;
                         bytes_left_to_write -= bytes_written;
+                } else if ((errno == ENOTSOCK)) {
+                        is_a_socket = 0;
                 } else if ((errno != EINTR)) {
                         break;
                 }
Index: plymouth-0.9.2/configure.ac
===================================================================
--- plymouth-0.9.2.orig/configure.ac
+++ plymouth-0.9.2/configure.ac
@@ -175,6 +175,8 @@ AC_DEFINE_UNQUOTED(BOOT_TTY, "$BOOT_TTY"
 AC_ARG_WITH(shutdown-tty, AS_HELP_STRING([--with-shutdown-tty=<tty>],[Default TTY to use in shutdown mode (by default tty63)]),SHUTDOWN_TTY=${withval},SHUTDOWN_TTY=/dev/tty63)
 AC_DEFINE_UNQUOTED(SHUTDOWN_TTY, "$SHUTDOWN_TTY", [TTY to use in shutdown mode])
 
+AC_CHECK_DECLS([MSG_NOSIGNAL], [], [], [[ #include <sys/socket.h> ]])
+
 # Turn on the additional warnings last, so -Werror doesn't affect other tests.
 
 AC_DEFUN([PLYMOUTH_CC_TRY_FLAG], [
